# **Exploratory Data Analysis** of DicodingCollection

___
**Rujukan atau proyek (saya) terkait**:

* [Data Wrangling](https://colab.research.google.com/drive/15QK3TOKyVhCBXQpSalcYsIxCg0jMoKTh?authuser=9#scrollTo=a5FsNPgo1ZuJ)

* [Exploratory Data Analysis](0)
___

## 1) Tahap Persiapan

In [1]:
# import library
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

In [4]:
# vscode load dataset
# define dictionary yang isinya (soon) name dataset
datasets = {
       'customers' : None,
       'orders' : None,
       'products' : None,
       'sales' : None
}

# looping untuk read dataset secara bersamaan
for name in datasets.keys():
    path = f"G:\My Drive\Dicoding\DicodingCollection\Sources\{name}.csv"
    datasets[name] = pd.read_csv(path)
    globals()[name] = datasets[name]
    print(f"{name} sukses dibaca") # If the print results can be output, it means dataaset sukses dibaca

customers sukses dibaca
orders sukses dibaca
products sukses dibaca
sales sukses dibaca


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# load dataset from google collab
# define dictionary yang isinya (soon) name dataseta
datasets = {
       'customers' : None,
       'orders' : None,
       'products' : None,
       'sales' : None
}

# looping untuk read dataset secara bersamaan
for name in datasets.keys():
    path = f"/content/drive/MyDrive/Dicoding/DicodingCollection/Sources/{name}.csv"
    datasets[name] = pd.read_csv(path)
    globals()[name] = datasets[name]
    print(f"{name} sukses dibaca") # If the print results can be output, it means dataaset sukses dibaca

customers sukses dibaca
orders sukses dibaca
products sukses dibaca
sales sukses dibaca


In [ ]:
customers

customer_id customer_name             gender   age  \
0              1       fulan 1             Female  30.0   
1              2       fulan 2  Prefer not to say  69.0   
2              3       fulan 3  Prefer not to say  59.0   
3              4       fulan 4  Prefer not to say  67.0   
4              5       fulan 5  Prefer not to say  30.0   
..           ...           ...                ...   ...   
995          996     fulan 996  Prefer not to say  59.0   
996          997     fulan 997  Prefer not to say  30.0   
997          998     fulan 998  Prefer not to say  32.0   
998          999     fulan 999  Prefer not to say  30.0   
999         1000    fulan 1000               Male  71.0   

                       home_address  zip_code                city  \
0    8606 Victoria TerraceSuite 560      5464       Johnstonhaven   
1        8327 Kirlin SummitApt. 461      8223     New Zacharyfort   
2         269 Gemma SummitSuite 109      5661            Aliburgh   
3         743 Bailey GroveSuite 141      1729   South Justinhaven   
4           48 Hyatt ManorSuite 375      4032      Griffithsshire   
..                              ...       ...                 ...   
995    0433 Armstrong HillSuite 974      7613     Lake Danielland   
996         04 Howell PassSuite 209      6950         Ellaborough   
997       72 Annabelle PassApt. 446        52          Kohlerberg   
998       170 Wilson AvenueApt. 577      7849      East Oscarfurt   
999      1671 Lauren KnollSuite 945      9012  Lake Audreyborough   

                            state    country  
0              Northern Territory  Australia  
1                 South Australia  Australia  
2    Australian Capital Territory  Australia  
3                      Queensland  Australia  
4                      Queensland  Australia  
..                            ...        ...  
995                      Tasmania  Australia  
996                      Tasmania  Australia  
997                    Queensland  Australia  
998             Western Australia  Australia  
999                      Tasmania  Australia  

[1000 rows x 9 columns]

In [ ]:
# Calling only one dataset = name
# Looping for all datasets = datasets[name]

In [ ]:
for name, data in datasets.items():
    print(f"{name} area ============")  # Menampilkan header dengan nama dataset
    print(data.describe())
    print("\n")

customers area ============
       customer_id          age     zip_code
count  1000.000000  1000.000000  1000.000000
mean    500.500000    49.860000  5004.872000
std     288.819436    17.647828  2884.497332
min       1.000000    20.000000     2.000000
25%     250.750000    34.000000  2401.750000
50%     500.500000    50.000000  5083.000000
75%     750.250000    65.000000  7460.250000
max    1000.000000    80.000000  9998.000000


orders area ============
          order_id  customer_id       payment
count  1000.000000  1000.000000   1000.000000
mean    500.500000   506.640000  33972.936000
std     288.819436   277.115502  14451.609047
min       1.000000     1.000000  10043.000000
25%     250.750000   275.250000  21329.250000
50%     500.500000   515.000000  33697.500000
75%     750.250000   737.250000  46249.000000
max    1000.000000  1000.000000  59910.000000


products area ============
        product_id        price     quantity
count  1260.000000  1260.000000  1260.000000
mean   

Sekilas, tidak ada yang aneh. Mari lihat informasi umumnya untuk melihat apakah terdapat perubahan pada dataset yang telah di wrangling sebelumnya.

In [ ]:
for name, data in datasets.items():
    print(f"{name} area ============")  # Menampilkan header dengan nama dataset
    print(data.info())
    print("\n")

customers area ============
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   customer_id    1000 non-null   int64  
 1   customer_name  1000 non-null   object 
 2   gender         1000 non-null   object 
 3   age            1000 non-null   float64
 4   home_address   1000 non-null   object 
 5   zip_code       1000 non-null   int64  
 6   city           1000 non-null   object 
 7   state          1000 non-null   object 
 8   country        1000 non-null   object 
dtypes: float64(1), int64(2), object(6)
memory usage: 70.4+ KB
None


orders area ============
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   order_id       1000 non-null   int64 
 1   customer_id    1000 non-null   int64 
 2   paymen

In [ ]:
for name, data in datasets.items():
    print(f"{name} area ============")  # Menampilkan header dengan nama dataset
    print(data.isnull().sum())
    print("\n")

customers area ============
customer_id      0
customer_name    0
gender           0
age              0
home_address     0
zip_code         0
city             0
state            0
country          0
dtype: int64


orders area ============
order_id         0
customer_id      0
payment          0
order_date       0
delivery_date    0
dtype: int64


products area ============
product_id      0
product_type    0
product_name    0
size            0
colour          0
price           0
quantity        0
description     0
dtype: int64


sales area ============
sales_id          0
order_id          0
product_id        0
price_per_unit    0
quantity          0
total_price       0
dtype: int64




In [ ]:
for name, data in datasets.items():
    print(f"{name}: {data.duplicated().sum()}")

customers: 0
orders: 0
products: 0
sales: 0


In [ ]:
customers[customers['age'] > 50]

customer_id customer_name             gender   age  \
1              2       fulan 2  Prefer not to say  69.0   
2              3       fulan 3  Prefer not to say  59.0   
3              4       fulan 4  Prefer not to say  67.0   
6              7       fulan 7  Prefer not to say  76.0   
7              8       fulan 8  Prefer not to say  75.0   
..           ...           ...                ...   ...   
988          989     fulan 989  Prefer not to say  53.0   
993          994     fulan 994             Female  53.0   
994          995     fulan 995  Prefer not to say  63.0   
995          996     fulan 996  Prefer not to say  59.0   
999         1000    fulan 1000               Male  71.0   

                      home_address  zip_code                city  \
1       8327 Kirlin SummitApt. 461      8223     New Zacharyfort   
2        269 Gemma SummitSuite 109      5661            Aliburgh   
3        743 Bailey GroveSuite 141      1729   South Justinhaven   
6      92 Hills Station StApt. 683       793           Masonfurt   
7       383 Muller SummitSuite 809      7681             Samside   
..                             ...       ...                 ...   
988       9003 Gemma RidgeApt. 954       698         Stewartland   
993       3128 Ivy CircleSuite 713       508           Hallshire   
994  19 Jordan Station StSuite 655      3742       Hermistonberg   
995   0433 Armstrong HillSuite 974      7613     Lake Danielland   
999     1671 Lauren KnollSuite 945      9012  Lake Audreyborough   

                            state    country  
1                 South Australia  Australia  
2    Australian Capital Territory  Australia  
3                      Queensland  Australia  
6                      Queensland  Australia  
7              Northern Territory  Australia  
..                            ...        ...  
988  Australian Capital Territory  Australia  
993             Western Australia  Australia  
994             Western Australia  Australia  
995                      Tasmania  Australia  
999                      Tasmania  Australia  

[492 rows x 9 columns]

Missing value dan duplicated value telah teratasi, tidak ada perubahan dengan proyek sebelumnya. Akan tetapi, terdapat perubahan dalam tipedata variabel order_date dan delivery, yang seharusnya datetime kembali berubah menjadi tipedata asal dari data yakni object. Hal itu bisa terjadi dikarenakan *csv tidak dapat menyimpan tipedata* dan karena hal itu, *harus dikonversi* setiap kali data ingin dibaca atau proses.

Sebagai catatan, CSV adalah format yang hanya menyimpan data sebagai teks. Ketika tipe data diubah menjadi datetime pada notebook sebelumnya, perubahan tersebut hanya berlaku di lingkungan notebook tersebut dan tidak tersimpan dalam file CSV [🔗](https://docs.python.org/3/library/csv.html).

Selain itu, nilai variabel age pada dataset customers juga mengalami perubahan tipedata menjadi float (desimal). Yang penyebabnya dapat ditemukan melalui proses eliminasi.

1. Data yang tidak konsisten
Hal ini dapat dipengaruhi oleh format csv yang dipakai untuk format penyimpanan dataset. Seperti yang dijelaskan di atas, csv tidak menyimpan tipedata. Melainkan hanya menyimpan data sebagai teks. Dan dari hal itu, (jadi) diketahui bahwa jika ada nilai dalam kolom age yang berupa teks atau memiliki karakter yang tidak sesuai dengan tipe data numerik (misalnya, ada tanda koma, spasi, atau karakter non-numerik lainnya). Yang membuat pandas (secara otomatis, mengambil asumsi) menganggap seluruh kolom sebagai tipe float.
2. Missing value (NaN)?
Tidak ditemukan missing value pada seluruh dataset yang ada, yang membuat missing value dikeluarkan dari proses eliminasi ini.

Mari lihat kode berikut:

In [ ]:
customers['age'].unique()

array([30., 69., 59., 67., 40., 76., 75., 51., 70., 39., 78., 42., 36.,
       34., 32., 79., 41., 55., 31., 20., 49., 50., 38., 44., 80., 27.,
       65., 37., 60., 68., 54., 56., 63., 52., 47., 62., 45., 33., 58.,
       64., 23., 48., 46., 53., 72., 66., 77., 73., 71., 35., 57., 28.,
       26., 22., 25., 29., 43., 24., 21., 61., 74.])

Sesuai dengan output yang dikeluarkan, penyebab perubahan tipe data pada variabel age ialah karena Pandas secara default akan mengonversi tipe data kolom menjadi float ... jika ada kemungkinan adanya angka desimal atau nilai yang tidak sesuai dengan tipe data integer, meskipun dalam banyak kasus data tersebut hanya mengandung nilai bulat.

In [5]:
customers['age'] = customers['age'].astype(int)
customers['age']

0      30
1      69
2      59
3      67
4      30
       ..
995    59
996    30
997    32
998    30
999    71
Name: age, Length: 1000, dtype: int32

In [ ]:
for name, data in datasets.items():
    print(f"{name} area ============")  # Menampilkan header dengan nama dataset
    print(data.head())
    print("\n")

customers area ============
   customer_id customer_name             gender  age  \
0            1       fulan 1             Female   30   
1            2       fulan 2  Prefer not to say   69   
2            3       fulan 3  Prefer not to say   59   
3            4       fulan 4  Prefer not to say   67   
4            5       fulan 5  Prefer not to say   30   

                     home_address  zip_code               city  \
0  8606 Victoria TerraceSuite 560      5464      Johnstonhaven   
1      8327 Kirlin SummitApt. 461      8223    New Zacharyfort   
2       269 Gemma SummitSuite 109      5661           Aliburgh   
3       743 Bailey GroveSuite 141      1729  South Justinhaven   
4         48 Hyatt ManorSuite 375      4032     Griffithsshire   

                          state    country  
0            Northern Territory  Australia  
1               South Australia  Australia  
2  Australian Capital Territory  Australia  
3                    Queensland  Australia  
4            

In [ ]:
for name, data in datasets.items():
    print(f"{name} area ============")  # Menampilkan header dengan nama dataset
    print(data.tail())
    print("\n")

customers area ============
     customer_id customer_name             gender  age  \
995          996     fulan 996  Prefer not to say   59   
996          997     fulan 997  Prefer not to say   30   
997          998     fulan 998  Prefer not to say   32   
998          999     fulan 999  Prefer not to say   30   
999         1000    fulan 1000               Male   71   

                     home_address  zip_code                city  \
995  0433 Armstrong HillSuite 974      7613     Lake Danielland   
996       04 Howell PassSuite 209      6950         Ellaborough   
997     72 Annabelle PassApt. 446        52          Kohlerberg   
998     170 Wilson AvenueApt. 577      7849      East Oscarfurt   
999    1671 Lauren KnollSuite 945      9012  Lake Audreyborough   

                 state    country  
995           Tasmania  Australia  
996           Tasmania  Australia  
997         Queensland  Australia  
998  Western Australia  Australia  
999           Tasmania  Australia  


or

Sekarang, adalah untuk mengetahui perubahan tipedata pada variabel order_date dan delivery_date pada dataset ordes.

In [ ]:
orders.sample(5)

order_id  customer_id  payment  order_date delivery_date
541       542          282    38737  2021-02-10    2021-02-14
9          10          659    21922  2021-10-15    2021-10-16
440       441          726    13419  2021-02-28    2021-03-02
270       271          469    39282  2021-06-07    2021-06-09
299       300          912    47255  2021-09-30    2021-10-25

Seperti sebelumnya, karena file yang digunakan adalah file berformat csv, maka perubahan tipedata akan kembali dilakukan.

In [ ]:
orders.dtypes

order_id          int64
customer_id       int64
payment           int64
order_date       object
delivery_date    object
dtype: object

In [6]:
d_col = ['order_date', 'delivery_date']

for x in d_col:
    orders[x] = pd.to_datetime(orders[x])

print(orders.dtypes)

order_id                  int64
customer_id               int64
payment                   int64
order_date       datetime64[ns]
delivery_date    datetime64[ns]
dtype: object


## 2) Tahap Business Inquiries

Pada tahap ini, tujuan dari bisnis DicodingCollection akan didefinisikan dengan seksama. Mulai dari tujuan bisnisnya hingga permasalahan yang perusahaan tersebut hadapi. Hal ini merupakan langkah yang sangat penting, karena tanpa pertanyaan yang jelas, eksplorasi data bisa menjadi tidak terarah.

### business understanding

#### langkah 1: Identifikasi Tujuan Bisnis

Tujuan bisnis ini berfokus pada apa yang ingin dicapai oleh perusahaan secara keseluruhan, yakni:

* Kinerja penjualan produk: *Bagaimana meningkatkan penjualan produk di DicodingCollection?*
* Produk fashion yang paling banyak diminati: *Produk fashion mana yang paling diminati oleh pelanggan?*
* Memahami referensi pelanggan: *Apa preferensi pelanggan terhadap produk yang di tawarkan?*
* Strategi campaign yang lebih efektif: *Bagaimana strategi campaign dapat dibuat lebih efektif untuk meningkatkan keterlibatan dan konversi?*

#### langkah 2: Terjemahkan Tujuan Bisnis ke Tujuan Data

Setelah tujuan bisnis ditetapkan, tujuan diubah menjadi sesuatu yang lebih spesifik yang bisa diukur dan dianalisis menggunakan data.

* Bagaimana cara meningkatkan penjualan produk di DicodingCollection:
 * faktor apa saja yang dapat mendorong peningkatan penjualan produk di DicodingCollection?
 * Bagaimana performa penjualan produk dan revenue dalam beberapa bulan terakhir?
* Produk fashion mana yang paling diminati oleh pelanggan:
 * Produk apa yang paling banyak dan sedikit terjual?
* Apa preferensi pelanggan terhadap produk yang di tawarkan:
 * Bagaimana distribusi demografi pelanggan yang dimiliki?
 * Seberapa sering seorang pelanggan melakukan pembelian pada beberapa bulan terakhir?
 * Kapan terakhir kali pelanggan melakukan transaksi?
* Bagaimana strategi campaign dapat dibuat lebih efektif untuk meningkatkan keterlibatan dan konversi:
 * Apa elemen-elemen campaign yang berpengaruh terhadap keterlibatan dan konversi pelanggan?
 * Apakah ada pola tertentu dalam jenis produk yang sering dibeli saat campaign sedang berlangsung?

### Langkah 3: Susun Pertanyaan Analitis

Dari dua tahapan sebelumnya, adapun pertanyaan analitis yang didapatkan;

* **Apakah terdapat hubungan antara total penjualan, kategori pembelian dan waktu pemesanan**
* **Apakah harga mempengaruhi total penjualan dan preferensi produk?**
* **Bagaimana tren penjualan  di beberapa bulan terakhir?**
* **Apakah terdapat pola dalam preferensi pelanggan terhadap pembelian produk?**
* **Bagaimana tingkat penjualan produk berdasarkan kategori (misalnya pakaian pria dan wanita)?**
* **Apakah terdapat perbedaan dalam perilaku pelanggan berdasarkan pembagian distribusi demografi (misalnya apakah pelanggan dengan usia tertentu memiliki kecenderungan terhadap produk tertentu)?**
* **Apakah ada kaitan antara jenis campaign yang diadakan dengan tingkat penjualan produk?**
* **Apakah keterlibatan pelanggan memiliki dampak yang signifikan terhadap konversi?**
* **Bagaimana pengaruh konversi yang dilakukan pelanggan sebelumnya dapat menarik keterlibatan pelanggan baru?**
* **Bagaimana pengaruh frekuensi campaign terhadap tingkat konversi pelanggan?**

## 3) Tahap EDA

Pada tahap ini, akan dilakukan ekplorasi data berdasarkan ertanyaan analitis yang sudah disusun.

#### customer

In [ ]:
customers.describe(include="all")

customer_id customer_name             gender          age  \
count   1000.000000          1000               1000  1000.000000   
unique          NaN          1000                  3          NaN   
top             NaN       fulan 1  Prefer not to say          NaN   
freq            NaN             1                742          NaN   
mean     500.500000           NaN                NaN    49.860000   
std      288.819436           NaN                NaN    17.647828   
min        1.000000           NaN                NaN    20.000000   
25%      250.750000           NaN                NaN    34.000000   
50%      500.500000           NaN                NaN    50.000000   
75%      750.250000           NaN                NaN    65.000000   
max     1000.000000           NaN                NaN    80.000000   

                          home_address     zip_code         city  \
count                             1000  1000.000000         1000   
unique                            1000          NaN          961   
top     8606 Victoria TerraceSuite 560          NaN  East Sophia   
freq                                 1          NaN            3   
mean                               NaN  5004.872000          NaN   
std                                NaN  2884.497332          NaN   
min                                NaN     2.000000          NaN   
25%                                NaN  2401.750000          NaN   
50%                                NaN  5083.000000          NaN   
75%                                NaN  7460.250000          NaN   
max                                NaN  9998.000000          NaN   

                  state    country  
count              1000       1000  
unique                8          1  
top     South Australia  Australia  
freq                139       1000  
mean                NaN        NaN  
std                 NaN        NaN  
min                 NaN        NaN  
25%                 NaN        NaN  
50%                 NaN        NaN  
75%                 NaN        NaN  
max                 NaN        NaN

Analisis:
* Variabel customer_name, home_adress dan city memiliki data kategori dengan jumlah nilai unik tertinggi, menunjukkan kemungkinan besar data ini tidak memiliki nilai yang sama di antara pelanggannya.
* Variabel age sebaliknya, memiliki variasi yang cukup beragam. Di mana berdasarkan batas usia normal yang telah ditentukan sebelumnya pada proses data wrangling, dapat jelaskan bahwa sebanyak 1000 pelanggan dalam rentang usia 20 - 80 tahun, dengan std. 17.64, merupakan segmentasi pelanggan yang dimiliki oleh DicodingCollection.
* Kategori gender menujukkan dominasi 'prefer not to say' sebanyak 742 pengulangan. Yang menunjukkan preferensi pelanggan terhadap privasi mereka.

Berdasarkan analisis sementara, variabel gender memiliki nilai unik yang lebih sederhana ketimbang variabel/columns lainnya. Dengan besarnya kategori seperti 'prefer not to say' jelas mengindikasikan adanya pembagian segmentasi di antara pelanggan. Segmentasi itulah yang pada dasarnya akan menjadi gambaran dari analisis preferensi pelanggan terhadap produk yang ada.

##### gender x age

Kemudian, untuk menganalisis apakah distribusi gender sejalan dengan distribusi terhadap usia pelanggan, maka dibuatlah pivot table untuk mencaritahu nilainya secara keseluruhan.

In [ ]:
customers.groupby(by="gender").agg({
    "customer_id" : "nunique",
    "age" : ["max", "min", "mean", "std"]
})

customer_id   age                            
                      nunique   max   min       mean        std
gender                                                         
Female                    115  79.0  20.0  49.147826  16.646607
Male                      143  80.0  20.0  51.230769  18.462635
Prefer not to say         742  80.0  20.0  49.706199  17.648386

Analisis:
* Seperti sebelumnya, prefer not to say mendominasi. Dengan keseluruhan pembagian gender yang ada, yakni rentang usia pelanggan di antara 20 hingga 80 tahun, sudah ada cukup bukti yang meyakinkan bahwa terdapat variasi dan segmentasi pasar dalam kelas gender dan age.

##### city n state

In [ ]:
customers.groupby(by="city").customer_id.nunique().sort_values(ascending=False)

city
East Aidan       3
East Sophia      3
New Ava          3
West Jackfort    2
Corkeryshire     2
                ..
Jordanside       1
Josephland       1
Josephmouth      1
Justinport       1
Zacville         1
Name: customer_id, Length: 961, dtype: int64

In [ ]:
customers.groupby(by="state").customer_id.nunique().sort_values(ascending=False)

state
South Australia                 139
Queensland                      134
New South Wales                 132
Northern Territory              125
Western Australia               124
Australian Capital Territory    121
Victoria                        121
Tasmania                        104
Name: customer_id, dtype: int64

Dapat dilihat bahwa persebaran yang ada pada tiap kota dan negara bagian cukup merata. Dengan detailnya seperti East Aidan, East Sophia dan New Ava adalah kota dengan jumlah pelanggan terbanyak dan banyaknya pelanggan yang berasal dari negara bagian South Australia. Informasi yang didapatkan sudah cukup untuk mengetahui mana market segment yang paling menjanjikan dalam penjualan di DicodingCollection.

### orders

In [ ]:
orders

order_id  customer_id  payment  order_date delivery_date
0           1           64    30811  2021-08-30    2021-09-24
1           2          473    50490  2021-02-03    2021-02-13
2           3          774    46763  2021-10-08    2021-11-03
3           4          433    39782  2021-05-06    2021-05-19
4           5          441    14719  2021-03-23    2021-03-24
..        ...          ...      ...         ...           ...
995       996          345    37843  2021-01-13    2021-02-02
996       997          346    53831  2021-01-18    2021-01-31
997       998          407    53308  2021-05-05    2021-05-21
998       999          428    31643  2021-06-15    2021-07-12
999      1000          896    27836  2021-04-07    2021-04-24

[1000 rows x 5 columns]

Pada tahapan ini, akan dibuat variabel baru yang nantinya akan digunakan untuk menampung informasi yang ditemukan untuk detail pesanan. Di mana dataset orders di sini berkaitan dengan pesanan pelanggan dan pastinya sedikit terkait dengan pengiriman produk.


Terdapat perbedaan waktu antara order_date dan delivery_date. Mengapa?

In [ ]:
delivery_time = orders["delivery_date"] - orders["order_date"]
delivery_time

0     25 days
1     10 days
2     26 days
3     13 days
4      1 days
        ...  
995   20 days
996   13 days
997   16 days
998   27 days
999   17 days
Length: 1000, dtype: timedelta64[ns]

Untuk menghitung waktu pengiriman, perlu dicari selisih antara tanggal produk di kirim dan tanggal pembuatan pemesanan. Pada kasus ini, dengan melakukan operasi pengurangan biasa selisih dalam bentuk hari sejatinya sudah didapatkan. Dengan menggunakan tipedata timedelta, yang merupakan representasi dari cara pandas terhadap durasi waktu, itu tidak sepenuhnya fleksibel terhadap analisis lain yang berkenaan dengan satuan konversi lain (seperti hari atau jam).

Karena itu, selisih hari yang telah di dapatkan, dalam hal ini, akan dikonversi menjadi satuan detik yang fleksibel terhadap satuan konversi lainnya.

Seperti contoh, jika mengambil index ke-0, di mana durasi pengiriman yang didapatkan ialah 25 hari. Jika dijabarkan:

In [ ]:
# satu menit itu 60 detik
# satu jam itu 60 menit
# dalam satu jam berarti ada 60 detik sebanyak 60.

# jika dilakukan operasi sederhana, maka:

a = 60  # detik dalam satu menit
b = 60  # menit dalam satu jam
c = a * b  # detik dalam satu jam
print(c)

# satu hari ada 24 jam, maka:
d = c * 24
print(d)

3600
86400


In [ ]:
delivery_time = orders["delivery_date"] - orders["order_date"]
delivery_time = delivery_time.apply(lambda x: x.total_seconds())
orders["delivery_time"] = round(delivery_time/86400)

In [ ]:
orders

order_id  customer_id  payment order_date delivery_date  delivery_time
0           1           64    30811 2021-08-30    2021-09-24           25.0
1           2          473    50490 2021-02-03    2021-02-13           10.0
2           3          774    46763 2021-10-08    2021-11-03           26.0
3           4          433    39782 2021-05-06    2021-05-19           13.0
4           5          441    14719 2021-03-23    2021-03-24            1.0
..        ...          ...      ...        ...           ...            ...
995       996          345    37843 2021-01-13    2021-02-02           20.0
996       997          346    53831 2021-01-18    2021-01-31           13.0
997       998          407    53308 2021-05-05    2021-05-21           16.0
998       999          428    31643 2021-06-15    2021-07-12           27.0
999      1000          896    27836 2021-04-07    2021-04-24           17.0

[1000 rows x 6 columns]

In [ ]:
orders.describe(include="all")

order_id  customer_id       payment                  order_date  \
count  1000.000000  1000.000000   1000.000000                        1000   
mean    500.500000   506.640000  33972.936000  2021-05-27 18:38:52.800000   
min       1.000000     1.000000  10043.000000         2021-01-01 00:00:00   
25%     250.750000   275.250000  21329.250000         2021-03-13 18:00:00   
50%     500.500000   515.000000  33697.500000         2021-05-27 12:00:00   
75%     750.250000   737.250000  46249.000000         2021-08-12 06:00:00   
max    1000.000000  1000.000000  59910.000000         2021-10-24 00:00:00   
std     288.819436   277.115502  14451.609047                         NaN   

             delivery_date  delivery_time  
count                 1000    1000.000000  
mean   2021-06-10 20:31:12      14.078000  
min    2021-01-03 00:00:00       1.000000  
25%    2021-03-28 00:00:00       8.000000  
50%    2021-06-11 12:00:00      14.000000  
75%    2021-08-24 06:00:00      21.000000  
max    2021-11-20 00:00:00      27.000000  
std                    NaN       7.707225

Dari hasil yang dikeluarkan, diketahui bahwa setengah pesanan sampai ke tujuan  dalam 14 hari. Dengan durasi paling singkat yakni 1 hari dan 27 hari sebagai durasi pengiriman terlama, ditambah standar deviasi sekitar 7.7, menandakan bahwa dalam durasi pengiriman produk/pesanan pun terdapat variasi yang beragam. Hal tersebut, rupanya berbeda-beda dan dari 3 quartile yang ada, terdapat pembagian pengantaran pesanan atau logistik. Seperti yang lebih dari 21 hari, yakni quartile 3, perlu dianalisis lebih lanjut untuk mengetahui apa yang menghambat pengiriman produk.

#### customers x orders

In [12]:
customer_id_orders = orders.customer_id.tolist()
customers['status'] = customers['customer_id'].apply(lambda x: "Active" if x in customer_id_orders else "Not Active")
customers

customer_id customer_name             gender  age  \
0              1       fulan 1             Female   30   
1              2       fulan 2  Prefer not to say   69   
2              3       fulan 3  Prefer not to say   59   
3              4       fulan 4  Prefer not to say   67   
4              5       fulan 5  Prefer not to say   30   
..           ...           ...                ...  ...   
995          996     fulan 996  Prefer not to say   59   
996          997     fulan 997  Prefer not to say   30   
997          998     fulan 998  Prefer not to say   32   
998          999     fulan 999  Prefer not to say   30   
999         1000    fulan 1000               Male   71   

                       home_address  zip_code                city  \
0    8606 Victoria TerraceSuite 560      5464       Johnstonhaven   
1        8327 Kirlin SummitApt. 461      8223     New Zacharyfort   
2         269 Gemma SummitSuite 109      5661            Aliburgh   
3         743 Bailey GroveSuite 141      1729   South Justinhaven   
4           48 Hyatt ManorSuite 375      4032      Griffithsshire   
..                              ...       ...                 ...   
995    0433 Armstrong HillSuite 974      7613     Lake Danielland   
996         04 Howell PassSuite 209      6950         Ellaborough   
997       72 Annabelle PassApt. 446        52          Kohlerberg   
998       170 Wilson AvenueApt. 577      7849      East Oscarfurt   
999      1671 Lauren KnollSuite 945      9012  Lake Audreyborough   

                            state    country      status  
0              Northern Territory  Australia      Active  
1                 South Australia  Australia  Not Active  
2    Australian Capital Territory  Australia  Not Active  
3                      Queensland  Australia  Not Active  
4                      Queensland  Australia  Not Active  
..                            ...        ...         ...  
995                      Tasmania  Australia      Active  
996                      Tasmania  Australia  Not Active  
997                    Queensland  Australia      Active  
998             Western Australia  Australia  Not Active  
999                      Tasmania  Australia      Active  

[1000 rows x 10 columns]

In [14]:
customers.groupby(by='status').customer_id.count()

status
Active        617
Not Active    383
Name: customer_id, dtype: int64

In [15]:
customers_orders = pd.merge(
       left = orders,
       right = customers,
       how = 'left',
       left_on = 'customer_id',
       right_on = 'customer_id'
)
customers_orders.head()

order_id  customer_id  payment order_date delivery_date customer_name  \
0         1           64    30811 2021-08-30    2021-09-24      fulan 64   
1         2          473    50490 2021-02-03    2021-02-13     fulan 473   
2         3          774    46763 2021-10-08    2021-11-03     fulan 774   
3         4          433    39782 2021-05-06    2021-05-19     fulan 433   
4         5          441    14719 2021-03-23    2021-03-24     fulan 441   

              gender  age                   home_address  zip_code  \
0  Prefer not to say   75      4927 Alice MeadowApt. 960      7787   
1               Male   61  531 Schmitt BoulevardApt. 010      1744   
2  Prefer not to say   34       2096 Wilson MewsApt. 714      8590   
3  Prefer not to say   46        5777 Mayer PassApt. 881      9728   
4  Prefer not to say   53   33 Richards JunctionApt. 478      7650   

                 city              state    country  status  
0      Sanfordborough    South Australia  Australia  Active  
1             Annaton    South Australia  Australia  Active  
2  West Jonathanshire           Tasmania  Australia  Active  
3    West Michaelport           Tasmania  Australia  Active  
4     South Rileyview  Western Australia  Australia  Active

##### order berdasarkan kota

In [ ]:
customers_orders.groupby(by='city').order_id.count().sort_values(ascending=False).reset_index().head(10)

city  order_id
0        Jordanside         6
1           New Ava         6
2         Lake Rose         5
3        O'keefeton         5
4          East Max         5
5  Port Hannahburgh         5
6          Rubyfort         5
7          West Kai         5
8     Andrewborough         4
9     Port Aaronton         4

Dari grouping yang dihasilkan, Jordanside dan New Ava adalah kota dengan pesanan terbanyak yakni sejumlah 6 pesanan.

##### orders berdasarkan state

In [ ]:
customers_orders.groupby(by='state').order_id.count().sort_values(ascending=False).reset_index()

state  order_id
0               South Australia       148
1                    Queensland       139
2             Western Australia       130
3               New South Wales       129
4  Australian Capital Territory       118
5                      Victoria       118
6                      Tasmania       112
7            Northern Territory       106

##### orders berdasarkan gender

In [ ]:
customers_orders.groupby(by='gender').order_id.nunique().sort_values(ascending=False)

gender
Prefer not to say    725
Female               139
Male                 136
Name: order_id, dtype: int64

##### Orders berdasarkan demografi

Pada data wrangling, batas usia normal dan pembagian segment usia pelanggan telah dibagi. Sebelumnya, terdapat lebih dari 5 segment berdasarkan rentang usia pelanggan, hal itu cukup berguna untuk memperjelas batas usia normal manusia. Namun, untuk analisis mendatang, dibutuhkan segmentasi yang lebih ringkas untuk menampung hasilnya secara keseluruhan.

Pada data wrangling yang digunakan adalah ini:
```
* Usia anak-anak: 0 - 12 tahun
* Usia remaja: 13 - 19 tahun
* Usia dewasa muda: 20 - 35 tahun #25% of data or q1
* Usia dewasa: 36 - 50 tahun # 50% of data or q2, median and mean (menujukkan data simetris dan ada ada variasi)
* Usia paruh baya: 51 - 70 tahun # 75% of data or q3
* Usia lanjut: 71 - 100 tahun
* Usia ekstrem: Di atas 100 tahun (dianggap tidak realistis)
```

Tetapi pada proses ini, yang akan digunakan adalah dengan mengambil berdasarkan nilai quartile-nya. Yakni:
```
Youth: < 24 tahun
Adults: 25 - 64 tahun
Senior: > 65 tahun
```

Pada pembagian ini juga diterapkan kategori usia berdasarkan istilah umum yang digunakan pada bidang produk fashion.

In [16]:
customers_orders['age_group'] = customers_orders.age.apply(lambda x: 'Youth' if x <= 24 else('Seniors' if x > 64 else 'Adults'))
customers_orders

# groupby(by='age').

order_id  customer_id  payment order_date delivery_date customer_name  \
0           1           64    30811 2021-08-30    2021-09-24      fulan 64   
1           2          473    50490 2021-02-03    2021-02-13     fulan 473   
2           3          774    46763 2021-10-08    2021-11-03     fulan 774   
3           4          433    39782 2021-05-06    2021-05-19     fulan 433   
4           5          441    14719 2021-03-23    2021-03-24     fulan 441   
..        ...          ...      ...        ...           ...           ...   
995       996          345    37843 2021-01-13    2021-02-02     fulan 345   
996       997          346    53831 2021-01-18    2021-01-31     fulan 346   
997       998          407    53308 2021-05-05    2021-05-21     fulan 407   
998       999          428    31643 2021-06-15    2021-07-12     fulan 428   
999      1000          896    27836 2021-04-07    2021-04-24     fulan 896   

                gender  age                   home_address  zip_code  \
0    Prefer not to say   75      4927 Alice MeadowApt. 960      7787   
1                 Male   61  531 Schmitt BoulevardApt. 010      1744   
2    Prefer not to say   34       2096 Wilson MewsApt. 714      8590   
3    Prefer not to say   46        5777 Mayer PassApt. 881      9728   
4    Prefer not to say   53   33 Richards JunctionApt. 478      7650   
..                 ...  ...                            ...       ...   
995  Prefer not to say   62      9182 Hermann MallApt. 890       799   
996  Prefer not to say   40    59 Bergstrom GroveSuite 260      6060   
997               Male   41      5233 Lara AvenueSuite 291      8697   
998  Prefer not to say   66      80 Hermiston WaySuite 403      1112   
999  Prefer not to say   67     985 Nguyen TerraceApt. 362      7737   

                   city                         state    country  status  \
0        Sanfordborough               South Australia  Australia  Active   
1               Annaton               South Australia  Australia  Active   
2    West Jonathanshire                      Tasmania  Australia  Active   
3      West Michaelport                      Tasmania  Australia  Active   
4       South Rileyview             Western Australia  Australia  Active   
..                  ...                           ...        ...     ...   
995          Audreyfurt                      Victoria  Australia  Active   
996           Olsonbury            Northern Territory  Australia  Active   
997       Hartmannshire                    Queensland  Australia  Active   
998      North Liamfort  Australian Capital Territory  Australia  Active   
999            Ryanberg               South Australia  Australia  Active   

    age_group  
0     Seniors  
1      Adults  
2      Adults  
3      Adults  
4      Adults  
..        ...  
995    Adults  
996    Adults  
997    Adults  
998   Seniors  
999   Seniors  

[1000 rows x 15 columns]

Sekarang untuk melihat persebaran dari tiap group usia pelanggan yang telah dibuat:

In [19]:
customers_orders.groupby(by='age_group').order_id.nunique().sort_values(ascending=False)

age_group
Adults     681
Seniors    226
Youth       93
Name: order_id, dtype: int64

Adults adalah segment terbesar yang dimiliki oleh DicodingCollection dari segi usia. Besarnya kategori adults ini menjelaskan bahwa lebih dari setengah pelanggan DicodingCollection adalah pangsa pasar yang menjanjikan. Dimana rentang usia pelanggan yang melakukan pembelian berada dalam rentang usia 25 hingga 64 tahun, yang mencakup orang dewasa produktif. Ini adalah kelompok yang kemungkinan besar memiliki daya beli yang lebih tinggi dan dapat menjadi fokus utama untuk penawaran produk atau layanan, terutama yang berhubungan dengan kebutuhan profesional dan keluarga.

#### products

In [21]:
products.describe(include='all')

product_id product_type product_name  size colour        price  \
count   1260.000000         1260         1260  1260   1260  1260.000000   
unique          NaN            3           35     8      7          NaN   
top             NaN        Shirt        Denim    XS    red          NaN   
freq            NaN          420           70   252    180          NaN   
mean     629.500000          NaN          NaN   NaN    NaN   105.805556   
std      363.874979          NaN          NaN   NaN    NaN     9.704423   
min        0.000000          NaN          NaN   NaN    NaN    90.000000   
25%      314.750000          NaN          NaN   NaN    NaN    95.750000   
50%      629.500000          NaN          NaN   NaN    NaN   108.500000   
75%      944.250000          NaN          NaN   NaN    NaN   114.000000   
max     1259.000000          NaN          NaN   NaN    NaN   119.000000   

           quantity                                   description  
count   1260.000000                                          1260  
unique          NaN                                          1260  
top             NaN  A red coloured, XS sized, Oxford Cloth Shirt  
freq            NaN                                             1  
mean      60.150000                                           NaN  
std       11.670573                                           NaN  
min       40.000000                                           NaN  
25%       50.000000                                           NaN  
50%       60.000000                                           NaN  
75%       70.000000                                           NaN  
max       80.000000                                           NaN

In [22]:
sales.describe(include='all')

sales_id     order_id   product_id  price_per_unit    quantity  \
count  5000.000000  5000.000000  5000.000000     5000.000000  5000.00000   
mean   2499.500000   503.038200   634.053200      103.501600     1.99240   
std    1443.520003   285.964418   363.255794        9.195004     0.80751   
min       0.000000     1.000000     1.000000       90.000000     1.00000   
25%    1249.750000   258.000000   323.000000       95.000000     1.00000   
50%    2499.500000   504.500000   635.000000      102.000000     2.00000   
75%    3749.250000   749.000000   951.000000      112.000000     3.00000   
max    4999.000000   999.000000  1259.000000      119.000000     3.00000   

       total_price  
count  5000.000000  
mean    206.360000  
std      86.357457  
min      90.000000  
25%     112.000000  
50%     204.000000  
75%     285.000000  
max     357.000000

Dengan melihat hasilnya, diketahui bahwa rentang harga barang yang ada pada DicodingCollection berkisar di antara 90 - 119 dolar, dengan produk paling mahal yang terjual yakni di harga 357 dolar dengan pembelian sebanyak 3 item dari kelas harga tertinggi, menunjukan adanya kecendrungan dan preferensi dalam tiap segment pesar. Bahwa dalam produk dengan harga termahal sekalipun memiliki target pasarnya masing-masing.

In [24]:
products.sort_values(by='price' , ascending=False)

product_id product_type product_name size  colour  price  quantity  \
698          698       Jacket        Parka    L  violet    119        53   
665          665       Jacket        Parka   XS     red    119        65   
692          692       Jacket        Parka    M  indigo    119        66   
693          693       Jacket        Parka    L  indigo    119        44   
694          694       Jacket        Parka   XL  indigo    119        78   
...          ...          ...          ...  ...     ...    ...       ...   
1219        1219     Trousers      Pleated   XL  indigo     90        45   
1220        1220     Trousers      Pleated   XS  violet     90        58   
1221        1221     Trousers      Pleated    S  violet     90        50   
1222        1222     Trousers      Pleated    M  violet     90        45   
593          593       Jacket       Bomber    L  violet     90        68   

                                        description  
698        A violet coloured, L sized, Parka Jacket  
665          A red coloured, XS sized, Parka Jacket  
692        A indigo coloured, M sized, Parka Jacket  
693        A indigo coloured, L sized, Parka Jacket  
694       A indigo coloured, XL sized, Parka Jacket  
...                                             ...  
1219  A indigo coloured, XL sized, Pleated Trousers  
1220  A violet coloured, XS sized, Pleated Trousers  
1221   A violet coloured, S sized, Pleated Trousers  
1222   A violet coloured, M sized, Pleated Trousers  
593       A violet coloured, L sized, Bomber Jacket  

[1260 rows x 8 columns]

Seperti yang terlihat, memang terdapat variasi dalam segi harga yang ditawarkan. Dengan beberapa Produk yang memiliki harga 119 dolar, walaupun itu harga termahal yang ditawarkan, meunnjukan memang terdapat daya beli yang tinggi dari pelanggan terhadap DicodingCollection. Mengindidikasikan peranan DiCo yang berhasil menjadi Online Fashion yang menjadi pilihan pelanggan.

##### product berdasarkan 